### ISSUE TITLE
Handing out open tasks and collecting ideas for tasks

### ISSUE THREAD

**Issue by Tomivix** [Tomivix/hpa#19]
Since it seems you guys unfortunately can't edit the new project (I was sure you will be able to do that as contributors), I'm opening this issue for the sake of handing out open tasks. Just comment below if you want to get assigned for one of them or post your own idea and we will discuss it together.

You can see current status of the project [here](https://github.com/Tomivix/hpa/projects/1?fullscreen=true).

**Comment by mrwasp** [comment]
To process inputted code (one of tasks) we need a method which change code text to elements of our HashMaps.

I can take this task, but I have some questions:
1) What about HashMap<int, String> orders? We will store there Strings like "AR 1,2" and convert it to relevant Functions at the time of executing? 
2) Do we divide memory cells into vars and orders ones? In GUI we have it in one, but in Engine it is divided (HashMaps vars and orders).

**Comment by Tomivix** [comment]
Sorry for the wait but I went to buy a pizza and then were eating it, so you know...
Anyways, regarding your questions @mrwasp :
1. As far as I know this matter hasn't been decided yet but my personal idea is something along this lines: (All of this is up to change, if you have better idea, feel free to implement it)
   - User inputs code for directives and orders
   - User clicks 'Build' button
   - Code gets parsed and debugged
   - If it's valid put it into HashMap orders else throw errors with some feedback
   - Once it has been built 'Run' button gets enabled
   - When clicked by user code from HashMap orders is run
   -  If user changes code after bulding it has to be build once again
2. Not sure if we really need it splitted in two, I'll leave it up to whoever ends up implementing it...

If you want to take up this task comment back in this issue and I'll add you up.

**Comment by mrwasp** [comment]
Yes, I can take this.
![bez tytulu](https://cloud.githubusercontent.com/assets/23056020/20032449/5d36b8ca-a38a-11e6-9963-3f8e1f6c1271.png)
That's the way that you see action of build button?

**Comment by Tomivix** [comment]
Hmm, you already did something... Great, if that's the case you're assigned for this task.
I'll update the project right now. Also take a look at pull request #20 if you haven't already.

**Comment by mrwasp** [comment]
My thoughts about tasks:
- Licence/wiki sounds like good task for @Asiek777 :)
- @Tomivix what do you mean 'simple graphics'? Icons for buttons "Build" etc. rather than texts? I worked a little with GIMP, so if it's needed...

And my proposal for new task (not urgent but worth to do)
Improve checking code - show error when:
1) in commands RM when label in directives doesn't exist
2) in commands JUMP when label doesn't exist
3) in commands RR/RM - if number of register is bigger than 15
4) warning(?) when user try to change 14 or 15 register
5) in commands RM with arguments like 'offset(register)' - probably at the build time we won't know if it's correct (then processing returns segmentation fault), but we can check if register number is correct
6) (advanced) error log - specify type of error and number of line

ad. 4) Real "HPA" use 14 and 15 register while processing to locate orders/variables, so maybe we should implement it to our project? Actually 14 and 15 register can be used like each other one.

**Comment by Tomivix** [comment]
Sorry for the wait @mrwasp . Your answers below:
- I agree assigning license/wiki for @Asiek777 seems like easiest and most reasonable choice but at the same time I wonder if he wouldn't be displeased with the fact that he does no coding at all?
- Yes, that's exactly what I meant. Preparing two (disabled and enabled versions) graphics (some simple symbol + styled text) for each button and then using them inside ImageIcon. I think that would look better than those default mundane buttons. **If you're interested in this task then comment back.**
- I'll suprise you - I already have this task prepared, just haven't added it to project yet, since the prerequisite (the task you're currently doing) wasn't finished. I wanted it be something along your lines, basically it would give you some feedback when it spots an invalid line after clicking 'Build'. And I do plan to make it 'advanced' as you say from the begging (type of error and line included).
- Cool idea. It would require refractoring the code a bit, but sure why not. Will add it to the project.

**Comment by Aicedosh** [comment]
Keep in mind that JButton with ImageIcon looks sort of like this: http://www.devmanuals.com/images/images1/247SwingButton.gif

**Comment by Tomivix** [comment]
Whoops, my bad. I meant something more like what is on [this page](http://docs.oracle.com/javafx/2/ui_controls/button.htm).

**Comment by Aicedosh** [comment]
That's JafaFX not Swing. Nonetheless custom images as buttons will work perfectly fine, just draw them on JPanels with mouse listeners

**Comment by Tomivix** [comment]
Yeah right, it says so on the right side of the page... Regardless of method of implementation we still would need two custom pictures for each button (enabled and disabled version), right?

**Comment by Aicedosh** [comment]
Yes, that's correct

**Comment by Tomivix** [comment]
Good. How about MouseOn version? Or does JPanel indicate somehow that it's hovered?

**Comment by Aicedosh** [comment]
That's totally what I'm using to edit EditableCell and change it's background

**Comment by Tomivix** [comment]
Ok, so one picture for disabled state, one or enabled plus changing background color will do the trick?
Or should we make a separate picture for hovered button?

**Comment by Aicedosh** [comment]
No need to checge color, just change the image being drawn. So yes, two images for all buttons

**Comment by Tomivix** [comment]
Okay but shouldn't we have image change slightly when the button is hovered? Wouldn't that make three pictures in total (disabled, enabled, hovered)?

**Comment by Aicedosh** [comment]
If we want the disable state on particular button then yes, three. But most of them won't need to be disabled I believe

**Comment by Tomivix** [comment]
Well, most won't, but for example 'Run' and 'Step' buttons should be disabled before successful build, right? Therefore I think we should get three pictures for each button. It will probably be just the regular picture grayed out so it shouldn't add much work.

**Comment by mrwasp** [comment]
I can do something in GIMP, summarizing we need run, build and step icons in disabled, enabled and hovered version.
Hovered means "when mouse is on it", yes?
Any ideas how they should look like? Similar to Eclipse's ones?

**Comment by Tomivix** [comment]
Yes, currently we need three icons for each of the three buttons which totals in nine icons. I'm also thinking of preparing some logo for our program. And yes, hovered means basically 'with mouse hovering over it'.
As for designs I have no ideas for now but I'll think it over tomorrow and give you some @mrwasp . I will also update project before going to sleep.

_(Interestingly enough, when I merged #20 it also automatically merged #21. Seems like GitHub detected that some of the commits came from another pull request and acted accordingly. Strange but works...)_

**Comment by Tomivix** [comment]
Project updated. If you care take a look at it [over here](https://github.com/Tomivix/hpa/projects/1?fullscreen=true).

**Comment by mrwasp** [comment]
![bez tytulu](https://cloud.githubusercontent.com/assets/23056020/20276704/aaddfc5e-aa9d-11e6-9042-b49a2e4af15a.png)

Something like this?

**Comment by Aicedosh** [comment]
I'd take it

**Comment by mrwasp** [comment]
You can from my branch 'wasp'.

**Comment by Aicedosh** [comment]
I'd ask for pause/stop button as well (you can see why on my newest pull request)

**Comment by mrwasp** [comment]
Yes, in a moment, I am testing right now

**Comment by mrwasp** [comment]
Pause added

**Comment by Aicedosh** [comment]
@Tomivix : You can add me wherever my name figures as suggested. However I think I won't be able to add anything relevant until Thursday or so.

**Comment by Tomivix** [comment]
Okay, as you wish @Aicedosh. You don't necessarily have to hurry, I just want to know who will take care of what to keep it all organised. Current deadline is ~~November 24th~~ ~~end of November~~ Christmas by the way.

**Comment by mrwasp** [comment]
I can take logo and 14th/15th register and probably something else after Thursday's exam.

**Comment by Tomivix** [comment]
Okay, will add you only for those two tasks for now @mrwasp. Comment once again after Thursday in case you want to be assigned for something else as well.
